### ISSUE TITLE
torch.sign() giving wrong gradients

### ISSUE THREAD

**Issue by stases** [pytorch/pytorch#74439]
### 🐛 Describe the bug

Hello, I recently came across a weird phenomenon when using torch.sign() in my loss function. Specifically, in the calculation of the loss function for a given vector v, I am calculating matrices of the form **S_ij = v_i - v_j**. In the code below, I made 2 versions of the loss function, and the only difference is:
```python
S_diff = torch.sign((labels[None, :] - labels[:, None]).squeeze().T)
S_diff = torch.sign((labels[:, None] - labels[None, :]).squeeze())
```
However, the two should be the same, as the transpose takes care of the different ordering between them. Below, I provide a small working example where I include both versions of the loss function, a toy neural network model and then for two equal networks I print the gradients of the final linear layer. What you can observe is that the gradients are not the same. I am unsure what is causing this issue. Although the difference might seem small, the trained models actually give very different final states (as if the one that uses the transpose doesn't learn well). Thank you in advance!
```python
import torch
import torch.nn as nn
import copy

# Loss version 1
def compute_lambda_i_version1(scores, labels):
    scores = scores.squeeze(1)
    z_diff = (scores[:, None] - scores[None, :]).squeeze()
    S_diff = torch.sign((labels[None, :] - labels[:, None]).squeeze().T) # Difference

    lambda_i =  (1 - S_diff)  / 2 - 1 /(1 + torch.exp(z_diff))
    lambda_i = lambda_i.sum(axis=1).unsqueeze(1)
        
    return lambda_i 

# Loss version 2
def compute_lambda_i_version2(scores, labels):
    scores = scores.squeeze(1)
    z_diff = (scores[:, None] - scores[None, :]).squeeze()
    S_diff = torch.sign((labels[:, None] - labels[None, :]).squeeze()) # Difference

    lambda_i =  (1 - S_diff)  / 2 - 1 /(1 + torch.exp(z_diff))

    lambda_i = lambda_i.sum(axis=1).unsqueeze(1)
        
    return lambda_i 

# Generic neural net
class NeuralModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.network = nn.Sequential(
                        nn.Linear(501,256),
                        nn.ReLU(),
                        nn.Linear(256,1)
                        )
    
    def forward(self, x):
        x = self.network(x)
        return x

# Random input
inputs = torch.rand(5, 501)
labels = torch.FloatTensor([1, 2, 3, 0, 4])

# Init the networks
net_1 = NeuralModule()
net_2 = NeuralModule()
net_2.load_state_dict(copy.deepcopy(net_1.state_dict()))

# Network 1 gradients / loss version 1
scores_1 = net_1(inputs)
loss_1 = compute_lambda_i_version1(scores_1, labels)
torch.autograd.backward(scores_1, loss_1)
grad_1 = net_1.network[2].weight.grad

# Network 2 gradients / loss version 2
scores_2 = net_2(inputs)
loss_2 = compute_lambda_i_version2(scores_2, labels)
torch.autograd.backward(scores_2, loss_2)
grad_2 = net_2.network[2].weight.grad

print(torch.equal(grad_1, grad_2))
print(torch.abs(grad_1-grad_2).mean())
```

### Versions

```
PyTorch version: 1.11.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.7.11 (default, Jul 27 2021, 09:42:29) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19041-SP0
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 471.96
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.11.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py37h196d8e1_0
[conda] mkl_fft                   1.3.0            py37h46781fe_0
[conda] mkl_random                1.1.1            py37h47e9c7a_0
[conda] numpy                     1.18.1           py37h93ca92e_0
[conda] numpy-base                1.18.1           py37hc3f5095_1
[conda] torch                     1.11.0                   pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7
